### PR TITLE
timm to pytorch conversion for vit model fix

### PR DESCRIPTION
**Pull request by staghado**
This PR fixes this issue #26219 with timm to PyTorch conversion. It removes the need for hard coded values for model dims by using the attributes of the timm model without needing the model name.

It does the following things :

- [x] Extract model dims from the timm model directly, no need for ifs
- [x] Decides whether the converted model will be a classification model or only a feature extractor using the num_classes attribute of the timm model.
- [x] In the case of a feature extractor only model : remove the pooling layers from the PyTorch model and compare the output to the last hidden state instead.

This works for a large number of models in the ViT family.

@ArthurZucker, @amyeroberts, @rwightman

### CODE REVIEW

File: src/transformers/models/vit/convert_vit_timm_to_pytorch.py [hunk 1]
```diff
@@ -60,13 +60,13 @@ def create_rename_keys(config, base_model=False):
...
-        # layernorm + pooler
+        # layernorm
         rename_keys.extend(
             [
                 ("norm.weight", "layernorm.weight"),
                 ("norm.bias", "layernorm.bias"),
-                ("pre_logits.fc.weight", "pooler.dense.weight"),
-                ("pre_logits.fc.bias", "pooler.dense.bias"),
+                # ("pre_logits.fc.weight", "pooler.dense.weight"),
+                # ("pre_logits.fc.bias", "pooler.dense.bias"),
```
Comment (ArthurZucker) [inline src/transformers/models/vit/convert_vit_timm_to_pytorch.py:69]: to remove?

File: src/transformers/models/vit/convert_vit_timm_to_pytorch.py [hunk 2]
```diff
@@ -219,8 +198,8 @@ def convert_vit_checkpoint(vit_name, pytorch_dump_folder_path):
 
     if base_model:
         timm_pooled_output = timm_model.forward_features(pixel_values)
-        assert timm_pooled_output.shape == outputs.pooler_output.shape
-        assert torch.allclose(timm_pooled_output, outputs.pooler_output, atol=1e-3)
+        assert timm_pooled_output.shape == outputs.last_hidden_state.shape
+        assert torch.allclose(timm_pooled_output, outputs.last_hidden_state, atol=1e-1)
```
Comment (ArthurZucker) [inline src/transformers/models/vit/convert_vit_timm_to_pytorch.py:202]: not sure if the last hidden states are always gonna have the same shape as the pooled outputs?

**Review by ArthurZucker — COMMENTED** [review]
Thanks for the cleanup, LGTM but I need a second look from @rwightman 🤗

**Review by ArthurZucker — DISMISSED** [review]

File: src/transformers/models/vit/convert_vit_timm_to_pytorch.py (hunk 2 above)
Comment (staghado) [reply src/transformers/models/vit/convert_vit_timm_to_pytorch.py:202]: The 'timm_pooled_output' is the output of the forward_features method ([L656](https://github.com/huggingface/pytorch-image-models/blob/e1e7cf5275600ea74d7b6ed80cc105701bddefdb/timm/models/vision_transformer.py#L656)) which does not include a pooling layer, so I don't know why it's called a pooled output. I might be wrong though

Comment (rwightman) [reply src/transformers/models/vit/convert_vit_timm_to_pytorch.py:202]: that is correct, forward_features() is unpooled output

File: src/transformers/models/vit/convert_vit_timm_to_pytorch.py (hunk 1 above)
Comment (rwightman) [reply src/transformers/models/vit/convert_vit_timm_to_pytorch.py:69]: I removed support for pre-logits from timm vits because they weren't being used, and was determined in ('How to Train your ViT' and later that they didn't help much). They were only present in some in21k pretrained (original) vits that had zero'ed out classifier heads..

File: src/transformers/models/vit/convert_vit_timm_to_pytorch.py [hunk 3]
```diff
@@ -140,60 +140,39 @@ def convert_vit_checkpoint(vit_name, pytorch_dump_folder_path):
...
+
+    # retrieve architecture-specific parameters from the timm model
+    config.hidden_size = timm_model.embed_dim
+    config.intermediate_size = timm_model.blocks[0].mlp.fc1.out_features
+    config.num_hidden_layers = len(timm_model.blocks)
+    config.num_attention_heads = timm_model.blocks[0].attn.num_heads
+
+    # check whether the model has a classification head or not
+    if timm_model.num_classes != 0:
+        config.num_labels = timm_model.num_classes
```
Comment (rwightman) [inline src/transformers/models/vit/convert_vit_timm_to_pytorch.py:160]: So, this block here assumes in1k is the only situation when num_classes != 0 which is certainly not true these days, it was back when vit was first released, only models with valid classifiers were 1k (21k were all zero'd). 

Now timm has following layouts for vits that I can see:
* 1000 - Imagenet-1k 
* 21843 - Google's ImageNet-21k layout, from How to Train Your ViT paper (differs from other standard 21/22k layouts with 21841 classes)
* 11821 - my 'ImageNet-12k' layout
* A few models from CLIP weights that have a final embedding projection as num_classes (512, 768, etc), should not map them to classes

So, to be correct and get the mapping, something like that needs to be used https://github.com/huggingface/api-inference-community/blob/main/docker_images/timm/app/pipelines/image_classification.py#L31-L33 

However, I will point out this reveals the flaw in the id2label/label2id scheme, it's not adequate to cover situation like ImageNet-21k when being used as a 2-way mapping that assumes uniquenes. https://huggingface.co/datasets/imagenet-1k/discussions/1#651c5d13aa89f7554c3e26ca

**Comment by rwightman** [comment]
Overall looking much better, should be more robust now.

A few things, there are a number of vit configurations supported in timm that are not, to my knowledge, supported in transformers. Should there be an attempt to detect? Thinking of some examples
* if `fc_norm` is present (norm after pooling)
* use of global average pooling in combination (or without) class token
* non-overlapping position and class token embedding
* CLIP style vit with .norm_pre layer present
* SigLIP style vit with .attn_pool layer present
* and soon, use of 'registers' via reg_token param

File: src/transformers/models/vit/convert_vit_timm_to_pytorch.py (hunk 3 above)
Comment (staghado) [reply src/transformers/models/vit/convert_vit_timm_to_pytorch.py:186]: I have tried to fix this by using the 'infer_imagenet_subset' function from timm.data. I have used the index_to_label_name() method from ImageNetInfo to get the label names. Let me know if this is not how it should be done

**Comment by ArthurZucker** [comment]
Hey! Make sure to rebase to only have your changes! 😉

**Comment by staghado** [comment]
> Hey! Make sure to rebase to only have your changes! 😉

I have reset the branch's history and leave only my changes which fix the issue here.

**Comment by staghado** [comment]
> Overall looking much better, should be more robust now.
> 
> A few things, there are a number of vit configurations supported in timm that are not, to my knowledge, supported in transformers. Should there be an attempt to detect? Thinking of some examples
> 
> * if `fc_norm` is present (norm after pooling)
> * use of global average pooling in combination (or without) class token
> * non-overlapping position and class token embedding
> * CLIP style vit with `norm_pre` layer present
> * SigLIP style vit with `attn_pool` layer present
> * and soon, use of 'registers' via `reg_token` param
> * use of layer scale in ViT model blocks

I have tried to add some checks before trying to convert the model from timm to huggingface.
 Checks to be added :
    1. non-overlapping position and class token embedding 
    2. use of 'registers' via `reg_token` param
    3. check when a model has a convolution feature extractor like ResNet50

I have tested the script on the pre-trained ViTs and only the following give errors:

- vit_base_r50_s16_224.orig_in21k (contains a resnet block)
- vit_base_r50_s16_384.orig_in21k_ft_in1k (contains a resnet block)
- vit_small_r26_s32_224.augreg_in21k
- vit_small_r26_s32_224.augreg_in21k_ft_in1k
- vit_small_r26_s32_384.augreg_in21k_ft_in1k
- vit_tiny_r_s16_p8_224.augreg_in21k
- vit_tiny_r_s16_p8_224.augreg_in21k_ft_in1k
- vit_tiny_r_s16_p8_384.augreg_in21k_ft_in1k

**Comment by rwightman** [comment]
@staghado looking good, those hybrid resnet-vit models should be possible to catch (see if below) with a meaningful error .. other than looks ready to go 

`if not isinstance(model.patch_embed, timm.layers.PatchEmbed)` ...

**Review by rwightman — APPROVED** [review]
looks good from the timm perspective

**Review by amyeroberts — APPROVED** [review]
Thanks for adding this and making the script more general! 

Just a small question on an outstanding to-do. Otherwise LGTM!

File: src/transformers/models/vit/convert_vit_timm_to_pytorch.py [hunk 4]
```diff
@@ -140,60 +137,70 @@ def convert_vit_checkpoint(vit_name, pytorch_dump_folder_path):
...
+    if not isinstance(getattr(timm_model.blocks[0], "ls1", None), torch.nn.Identity) or not isinstance(
+        getattr(timm_model.blocks[0], "ls2", None), torch.nn.Identity
+    ):
+        raise ValueError(f"{vit_name} is not supported in transformers because it uses a layer scale in its blocks.")
+
+    # Hybrid ResNet-ViTs
+    if not isinstance(timm_model.patch_embed, timm.layers.PatchEmbed):
+        raise ValueError(f"{vit_name} is not supported in transformers because it is a hybrid ResNet-ViT.")
+
+    # non-overlapping position and class token embedding (to be added)
```
[the remark this answers is not in the document]
Comment (staghado) [reply src/transformers/models/vit/convert_vit_timm_to_pytorch.py:176]: Yes, but I am not sure how to do it, maybe check if:
`torch.allclose(timm_model.cls_token[0, 0, :], timm_model.pos_embed[0, 0, :], atol=1e-3)`?

[the remark this answers is not in the document]
Comment (rwightman) [reply src/transformers/models/vit/convert_vit_timm_to_pytorch.py:176]: right, that case, you can check an attribute on the timm model, `timm.model.no_embed_class` will be set to True when the cls embed doesn't overlap, although I think most models that use this would be caught by the layer scale check..

you can actually 'fix' this by prepending a zero value to the position embedding during conversion

[the remark this answers is not in the document]
Comment (staghado) [reply src/transformers/models/vit/convert_vit_timm_to_pytorch.py:176]: On all the `timm.list_pretrained('vit*')` models, the case when cls_embed does not overlap is not reached.
We can maybe add the check inside the layer scale case to have a fine-grained error message?

[the remark this answers is not in the document]
Comment (rwightman) [reply src/transformers/models/vit/convert_vit_timm_to_pytorch.py:176]: @staghado it's used for deit3 (again w/ layer scale). It will also be used in a few new experiments coming out soon related to registers and global average pool pooling, so, would also be incompatible for other reasons... so can't think of any model that'd use no_embed_class without also using another unsupported feature...

[the remark this answers is not in the document]
Comment (staghado) [reply src/transformers/models/vit/convert_vit_timm_to_pytorch.py:176]: @rwightman Ok, I will remove it for now!

**Comment by staghado** [comment]
@ArthurZucker

**Review by ArthurZucker — APPROVED** [review]
Thanks for improving this script! 🚀
